cpp/include/raft/neighbors/detail/nn_descent.cuh

```diff
@@ -1405,6 +1422,12 @@ void build(raft::resources const& res,
   auto int_graph = raft::make_host_matrix<int, int64_t, row_major>(
```

Comment:
I just realized that `int_graph` is also an excessive allocation... you could just copy the values directly from `graph_shrink_buffer` to the indices array allocated in the index object. You would need an extra template parameter on `GNND::build()` function to make that work. The NN Descent impl can only support `int` but RAFT aims to support `uint32_t`, `int8` and `uint8_t`. Which is fine, let NN Descent solve the problem in `int` and then copy from `graph_shrink_buffer` to the output indices array from the index in the correct type format.

Reply:
After looking into the code a bit, I think `int_graph` is needed.

As soon as `GNND::build()` is called, the given `output_graph` is casted like 
`graph_.h_graph      = (InternalID_t<Index_t>*)output_graph` where `InternalID_t<Index_t>*` ends up just being `int*` type. 

`InternalID_t<Index_t>` has to be an `int`(or at least a signed value) because of the way they keep track of old and new values (using the msb).

So passing the `idx.graph().data_handle()` directly to `GNND::build()` without allocating `int_graph` will end up having to change the logic inside `InternalID_t<Index_t>` because `idx.graph().data_handle()` could have types other than `int` (e.g. `uint32_t`).

Reply:
Ah you are right, I see there is a comment for that as well https://github.com/rapidsai/raft/blob/8fbce933bff14deee582eab345fb768f35cf7f27/cpp/include/raft/neighbors/detail/nn_descent.cuh#L994-L995